### PR TITLE
[ADP-2367] construct transaction info from CBOR encoded txs

### PR DESCRIPTION
**Pull request by paolino**
- abstract over transactions decorations to cover CBOR encoded ones
- add new Read.Tx features
   - inputs
   - collateral inputs
   - outputs
   - collateral outputs
   - fee
   - withdrawals
   - metadata
   - script validity
 - implement mkTransactionInfo for the  Read.Tx (CBOR encoded txs)
 
ADP-2367

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Features/Outputs.hs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import qualified Cardano.Wallet.Primitive.Types.Coin as Coin
+import qualified Cardano.Wallet.Primitive.Types.Hash as W
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Cardano.Wallet.Primitive.Types.TokenPolicy as W
+import qualified Cardano.Wallet.Primitive.Types.TokenQuantity as W
+import qualified Cardano.Wallet.Primitive.Types.Tx.TxOut as W
+
+getOutputs :: EraFun Outputs (K [W.TxOut])
+getOutputs = EraFun
+    { byronFun = \(Outputs ins) -> K . fmap BYC.fromTxOut $ toList ins
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Features/Outputs.hs]:
```suggestion
    { byronFun = \(Outputs os) -> K . fmap BYC.fromTxOut $ toList os
```

In is out. 🤪

File: lib/wallet/src/Cardano/Wallet/Read/Tx/CollateralInputs.hs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        , allegraFun = \_ -> CollateralInputs ()
+        , maryFun = \_ -> CollateralInputs ()
+        , alonzoFun = onTx $ \(AL.ValidatedTx b _ _ _) -> getCollateralInputs b
+        , babbageFun = onTx $ \(AL.ValidatedTx b _ _ _) -> getCollateralInputs b
+        }
+
+getCollateralInputs
+    :: ( HasField "collateral" a (CollateralInputsType b))
+    => a -> CollateralInputs b
+getCollateralInputs =  CollateralInputs . getField @"collateral"
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Read/Tx/CollateralInputs.hs:69]: The field name "collateral" is confusing — I checked again that this is indeed correct. 👍

File: lib/wallet/src/Cardano/Wallet/Read/Tx/CollateralOutputs.hs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        , maryFun = \_ -> CollateralOutputs ()
+        , alonzoFun = \_ -> CollateralOutputs ()
+        , babbageFun = onTx
+            $ \(AL.ValidatedTx b _ _ _) -> getCollateralOutputs b
+        }
+
+getCollateralOutputs
+    :: ( HasField "collateralReturn" a (CollateralOutputsType b))
+    => a -> CollateralOutputs b
+getCollateralOutputs =  CollateralOutputs . getField @"collateralReturn"
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Read/Tx/CollateralOutputs.hs:71]: The field name "collateralReturn" is confusing — I checked again that this is indeed correct. 👍

File: lib/wallet/src/Cardano/Wallet/DB/Store/Transactions/TransactionInfo.hs [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        , WT.txInfoTime = txTime
+        , WT.txInfoScriptValidity
+            = value $ getScriptValidity . getEraScriptValidity
+        }
+  where
+    tipH = getQuantity $ tip ^. #blockHeight
+    height = getQuantity submissionMetaHeight
+
+    txField :: EraFun Read.Tx (K a) -> a
+    txField f = extractEraValue $ applyEraFun f tx
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Store/Transactions/TransactionInfo.hs]:
```suggestion
    txField :: EraFun Read.Tx (K a) -> a
    txField = value
```

Uh — it's the same function? 🤔

**Review by HeinrichApfelmus — APPROVED** [review]
Excellent, thank you! 🥳 Only comment: The helper functions `value` and `txField` are exact duplicates?

At some point, we may want to pick one `TxIn` type that is valid for all eras and export it together in the upper levels of the hierarchy together with together `getInputs . getEraInputs`, and similar for other types where such a viewpoint is possible — but not today.

File: lib/wallet/src/Cardano/Wallet/DB/Store/Transactions/TransactionInfo.hs (hunk 4 above)
Comment (paolino) [reply lib/wallet/src/Cardano/Wallet/DB/Store/Transactions/TransactionInfo.hs]: I discovered a new way of documenting functions :shrug:

**Comment by paolino** [comment]
> At some point, we may want to pick one `TxIn` type that is valid for all eras and export it together in the upper levels of the hierarchy together with together `getInputs . getEraInputs`, and similar for other types where such a viewpoint is possible — but not today.

In theory, in the future, you could have a TxIn type that depends on 2 Reads primitive, so I think having access to the EraFun definition of things is not so bad. My crux is more like this: do we have to define top-level helpers for the compositions to leverage sharing of the composed records, or is the compiler smart enough?

**Pull request merged**